CPP-5096 S1236 Explain covered cases and exception (#3925)
SonarSource/rspec: rules/S1236/cfamily/rule.adoc

Copy assignment operators and move assignment operators can return anything, including ``++void++``.

However, if you decide to declare them yourself (don't forget the "Rule-of-Zero", S4963), 
it is a recommended practice to return a non-const reference to the assigned object (left-operand).
It allows the developer to chain the assignment operations, increasing consistency with what other types do and, in some cases, enabling the writing of concise code.

This rule will raise for assignment operators that deviate from the above expectation.

> === Using unconventional return type

Reviewer: ```suggestion
=== Using an unconventional return type
```

This rule will raise an issue if the return type of the copy or move assignment operator,
is different from mutable reference to the class type.

==== Noncompliant code example

[source,cpp,diff-id=1,diff-type=noncompliant]
----
class Clazz {
public:
  const Clazz& operator=(const Clazz& other); // Noncompliant, returns const reference
  Clazz operator=(Clazz&& other) noexcept;    // Noncompliant, returns by value
};
----

==== Compliant solution

[source,cpp,diff-id=1,diff-type=compliant]
----
class Clazz {
public:
  Clazz& operator=(const Clazz& other);      // Compliant
  Clazz& operator=(Clazz&& other) noexcept;  // Compliant
};
----

=== Returning object different that ``++*this++``

An assignment operator should return a reference to the assigned object.
Conventionally, such return is expressed as ``++return *this++``, and the rule will mark any return statement as deviating from this convention.

==== Noncompliant code example

[source,cpp,diff-id=2,diff-type=noncompliant]
----
class Clazz {
public:
  Clazz& set(Clazz& other);
  Clazz& operator=(Clazz const& other) {

>     return set(other); // Noncompliant, depends on return of `set` member function 

Reviewer: ```suggestion
    return set(other); // Noncompliant: depends on return of `set` member function 
```

  }

  Clazz&& operator=(Clazz&& other) noexcept {
    return other;      // Noncompliant, also return type is non-compliant
  }
};
----

==== Compliant solution

[source,cpp,diff-id=2,diff-type=compliant]
----
class Clazz {
public:
  Clazz& set(Clazz& other);
  Clazz& operator=(Clazz const& other) {
    set(other);
    return *this;      // Compliant
  }

  Clazz& operator=(Clazz&& other) noexcept {
    return *this;      // Compliant
  }
};
----

In {cpp}23, if the assignment operator is declared using an explicit object argument,
the rule will mark any return statement that does not return the object parameter directly.

==== Noncompliant code example

[source,cpp,diff-id=3,diff-type=noncompliant]
----
class Clazz {
public:
  Clazz& set(Clazz& other);
  Clazz& operator=(this Clazz& self, Clazz const& other) {

>     return self.set(other); // Noncompliant, depends on the return of `set` member function 

Reviewer: ```suggestion
    return self.set(other); // Noncompliant: depends on the return of `set` member function 
```

  }
};
----

==== Compliant solution

[source,cpp,diff-id=3,diff-type=compliant]
----
class Clazz {
public:
  Clazz& set(Clazz& other);
  Clazz& operator=(this Clazz& self, Clazz const& other) {
    self.set(other);
    return self;      // Compliant
  }
};
----

=== Declaring assignment operation as non-mutating

The assignment operation is designed to change the value of the target object,
to the same one as the source.

> Such operations are mutating and thus should not be declared with a `const` qualifier.

Reviewer: ```suggestion
Such operation is mutating and thus should not be declared with a `const` qualifier.
```

==== Noncompliant code example

[source,cpp,diff-id=4,diff-type=noncompliant]
----
class Clazz {
public:

>   Clazz& operator=(Clazz const& other) const { // Noncompliant, also least to noncompliant return statement

Reviewer: ```suggestion
  Clazz& operator=(Clazz const& other) const { // Noncompliant, also leads to noncompliant return statement
```

>   Clazz& operator=(Clazz const& other) const { // Noncompliant, also leads to noncompliant return statement

Reviewer: ```suggestion
  Clazz& operator=(Clazz const& other) const { // Noncompliant: also leads to noncompliant return statement
```

    return const_cast<Clazz&>(*this);
  }
  Class& operator=(Clazz&& other) const; // Noncompliant
};
----

==== Compliant solution

[source,cpp,diff-id=4,diff-type=compliant]
----
class Clazz {
public:

>   Clazz& operator=(Clazz const& other) const { // Compliant

Reviewer: ```suggestion
  Clazz& operator=(Clazz const& other) { // Compliant
```

    return *this;
  }

>   Clazz& operator=(Clazz&& other) const; // Compliant

Reviewer: ```suggestion
  Clazz& operator=(Clazz&& other); // Compliant
```

};
----

> When declaring assignment operator with {cpp}23 explicit object argument, 

Reviewer: ```suggestion
When declaring an assignment operator with {cpp}23 explicit object argument, 
```

> an assignment operator can be declared as non-mutation by either passing object argument:

Reviewer: ``` suggestion
the object argument should not be passed:
```

* by const reference - this is equivalent to declaring the implicit object parameter method as `const`,
                       as described above;
* by value - in this case a temporary object will be created, and modified by the assignment operator,
             instead of the left-hand side of the assignment operator

==== Noncompliant code example

[source,cpp,diff-id=5,diff-type=noncompliant]
----
class Clazz {
  int val;
public:

>   Clazz& operator=(this Clazz const& self, Clazz const& other) const { // Noncompliant, also leads to non-compliant return

Reviewer: ```suggestion
  Clazz& operator=(this Clazz const& self, Clazz const& other) const { // Noncompliant: also leads to non-compliant return
```
